tests/test_package_manager/test_v1.py

```diff
@@ -105,6 +126,59 @@ def test_sync(
             pm.sync(dev=True, third_party=False)
             update_patch.assert_called_with(package_id=DUMMY_PACKAGE_ID)
 
+        # test package already exists.
+        with mock.patch.object(pm, "add_package") as update_patch, mock.patch(
+            "pathlib.Path.exists", return_value=True
+        ), mock.patch(
+            "aea.helpers.ipfs.base.IPFSHashOnly.get", return_value=DUMMY_PACKAGE_HASH
+        ):
+            pm.sync(dev=True, third_party=False)
+            update_patch.assert_not_called()
+
+        # test package already exists but hash differs.
+        with mock.patch.object(pm, "add_package") as add_patch, mock.patch(
+            "pathlib.Path.exists", return_value=True
+        ), mock.patch(
+            "aea.helpers.ipfs.base.IPFSHashOnly.get",
+            return_value=DUMMY_PACKAGE_HASH + "X",
+        ):
+            with pytest.raises(PackageHashDoesNotMatch):
+                pm.sync(dev=True, third_party=False)
+            add_patch.assert_not_called()
+
+        # update hashes
+        pm = PackageManagerV1(
+            path=self.packages_dir_path,
+            dev_packages=OrderedDict({DUMMY_PACKAGE_ID: DUMMY_PACKAGE_HASH}),
+        )
+
+        with mock.patch.object(pm, "add_package") as add_patch, mock.patch(
+            "pathlib.Path.exists", return_value=True
+        ), mock.patch(
+            "aea.helpers.ipfs.base.IPFSHashOnly.get",
+            return_value=DUMMY_PACKAGE_HASH + "X",
+        ):
+            pm.sync(dev=True, third_party=False, update_hashes=True)
+            add_patch.assert_not_called()
+
+        # update packages
+        pm = PackageManagerV1(
+            path=self.packages_dir_path,
+            dev_packages=OrderedDict({DUMMY_PACKAGE_ID: DUMMY_PACKAGE_HASH}),
+        )
+
+        with mock.patch.object(pm, "add_package") as add_patch, mock.patch.object(
+            pm, "_update_package"
+        ) as update_patch, mock.patch(
+            "pathlib.Path.exists", return_value=True
+        ), mock.patch(
+            "aea.helpers.ipfs.base.IPFSHashOnly.get",
+            return_value=DUMMY_PACKAGE_HASH + "X",
+        ):
+            pm.sync(dev=True, third_party=False, update_packages=True)
+            update_patch.assert_called_once()
```

Comment:
there is a tiny context manager stacking tool, but it's not quite optimal. We can wrap it to make it nice so we can stack context managers in a way code is actually readable. 
https://github.com/valory-xyz/open-aea/blob/6e6a444ce2fee62932031f1bffdae58f24f52aeb/aea/test_tools/utils.py#L67-L72

I will add a tiny bit to this soon:tm: 
1. `yield stack`
2. we can then access stacked context managers by index (e.g. `stack.args[0]`). I want to wrap it so that we can initialize it with kwargs. E.g. 
```
with as_context(
    a=..., 
    b=...,
) as m: 
    ...
    m.a.assert_called()
    m.b.assert_not_called()
```

@0xArdi what do you think?

Reply:
interesting, but suppose can just reformat this block of `with ... as ...` better and make it more readable